### ISSUE TITLE
Align VSTS Tasks NuGet Restore, Packager, and Publisher to Select NuGet version - include version 4

### ISSUE THREAD

**Issue by KZeronimo** [Microsoft/vsts-tasks#3756]
Now that VS 2017, .Net Core Tools 1.0, and NuGet 4.0 are released the appropriate build tasks need to be bumped to leverage

An open question is will NuGet Restore be necessary for new solutions/projects with the availability of [background restore and msbuild /t:restore](http://blog.nuget.org/20161121/introducing-nuget4.0.html)

**Comment by alexmullans** [comment]
We're planning this work now, using the upcoming [tool installers](https://github.com/Microsoft/vsts-tasks/blob/master/docs/tools.md) to decouple version management from all the NuGet tasks.

In the meantime, here's a workaround if you need NuGet 4 immediately.

1. Add a PowerShell Script step before your NuGet steps. Choose the Inline Script option and enter Invoke-WebRequest https://dist.nuget.org/win-x86-commandline/v4.0.0/nuget.exe -OutFile $(build.artifactstagingdirectory)\nuget.exe

2. In your NuGet steps, set Path to NuGet.exe (in the Advanced section) to $(build.artifactstagingdirectory)\nuget.exe

**Comment by KZeronimo** [comment]
Perfect - thank you!

Taking the exact same approach with getting the release version of the .NET Core CLI tools installed on the hosted agent

Feel free to close

**Comment by leftler** [comment]
Here is a little more complicated script that still fits within the space requirements of a inline script. It will only download the 4.0 nuget if it detects that 4.0 or newer has not been installed yet.

```
if (-not (Test-Path $(Build.BinariesDirectory)\nuget.exe))
{
    $doDownload = $true
}
else
{
    $nuGetVersion = Get-ChildItem D:\Tools\nuget.exe | Select-Object -ExpandProperty VersionInfo | Select-Object -ExpandProperty ProductVersion
    if ([System.Version]$nuGetVersion -lt [System.Version]"4.0.0")
    {
        $doDownload = $true
    }
}
if($doDownload)
{
    Invoke-WebRequest https://dist.nuget.org/win-x86-commandline/v4.0.0/nuget.exe -OutFile $(Build.BinariesDirectory)\nuget.exe
}
```

Note this script uses `$(Build.BinariesDirectory)\nuget.exe` instead of `$(build.artifactstagingdirectory)\nuget.exe` for the custom NuGet path as that folder would be more approprate for downloading a binary tool like nuget that may get reused between builds.

**Comment by leftler** [comment]
Here is a slightly more complex version that still fits within the space limits of a inline script

```
if (-not (Test-Path $(Build.BinariesDirectory)\nuget.exe))
{
    $doDownload = $true
}
else
{
    $nuGetVersion = Get-ChildItem $(Build.BinariesDirectory)\nuget.exe | %{$_.VersionInfo} | %{$_.ProductVersion}
    if ([System.Version]$nuGetVersion -lt [System.Version]"4.0.0")
    {
        $doDownload = $true
    }
}
if($doDownload)
{
    Invoke-WebRequest https://dist.nuget.org/win-x86-commandline/v4.0.0/nuget.exe -OutFile $(Build.BinariesDirectory)\nuget.exe
}
```

This script checks to see if nuget.exe already exists and is version 4.0.0 or newer. if it does not exist or is a old version it downloads it.

Note this saves to `$(Build.BinariesDirectory)\nuget.exe` instead of `$(build.artifactstagingdirectory)\nuget.exe` as the binaries folder is not cleaned between builds and is where downloaded tools should be put.

**Comment by DanJ210** [comment]
Hey all. Really appreciate the work. I am new to all of this but I can no longer run .NET Core builds in BitBucket anymore. I get the error: "Can't find project.json" as obviously that has been replaced by .csproj. Is this something I have to fix or that BitBucket has to change? Below is my very simple and basic .yml file. Thanks in advance for any guidance, I've been searching a lot.

`# This is a sample build configuration for .NET Core.
# Check our guides at https://confluence.atlassian.com/x/5Q4SMw for more examples.
# Only use spaces to indent your .yml configuration.
# -----
# You can specify a custom docker image from Docker Hub as your build environment.
image: microsoft/dotnet:onbuild

pipelines:
  default:
    - step:
        script: # Modify the commands below to build your repository.
          - dotnet restore
          - dotnet build
#          - dotnet test TestName`

**Comment by marvinpuethe** [comment]
> Not an agent issue. It's the nuget task that's offering v4. Tasks are in that repo.
> The packaging team just checked that in so it should in the VSTS next rollout
> Microsoft/vsts-tasks@3e3ec62

see #873 5 days ago

For all those waiting ;)

**Comment by DanJ210** [comment]
@marvinpuethe In deeper investigation from VSTS I saw that there are version check boxes in the nuget package area in settings of the builds and saw there was not a 4.0. I did turn it off entirely but I still get a bunch of errors. So do you think that the above issue fix being rolled out will fix that entirely?

My other question would be if this affects bitbucket at all or is this something that bitbucket would have to fix as well? Would ne nice if it could be fixed from a VS2017 update but I'm not sure how that works, I'm just sayin.

Thanks for any responses.

**Comment by marvinpuethe** [comment]
@DanJ210 Sorry but I'm not participating in the development. Maybe @bryanmacfarlane could help you out.

**Comment by bryanmacfarlane** [comment]
@alexmullans is from the packaging team doing the work.

**Comment by alexmullans** [comment]
I can't speak to BitBucket at all - as far as I know, nothing in this library of VSTS build tasks relates to BitBucket.

For VSTS, we have an update going out in the current deployment to make NuGet 4 available as an option in the NuGet Installer task. We'll have a more comprehensive update with support for the new tool installer framework next sprint, but the current deployment should get you unblocked.

**Comment by DanJ210** [comment]
Thanks for your response @alexmullans. I was waiting for the update before I responded but after the VS2017 update and creating a new project out of the ASP.NET Core templates, builds are still failing on VTFS. Using Hosted 2017 as provided. What's crazy is that the first error is failing because project.json cannot be found. I thought this was getting changed?

Sorry for my newbieness but I guess there's an update slated for the actual VTFS site that allows me to select nuget package 4.0, I don't see that option as of now so that's the next thing I'm waiting for. Until that happens I can't publish to azure after builds so I'm a little stressed.

Thanks for all you guys/girls do.

**Comment by alexmullans** [comment]
@DanJ210 the option for NuGet 4 is available now - look in the Advanced section of the NuGet Installer task. I saw a PR out to make that the default for the ASP.NET (PREVIEW) template, but not sure that made the current deployment.

**Comment by alexmullans** [comment]
Duplicate of #3756

**Comment by DanJ210** [comment]
@alexmullans Hey alex. I appreciate the help. I did see the NuGet 4 selection when creating or editing a VTFS build and yes that does solve the NuGet problem which is awesome! I still can't get builds to pass. I'm using projects that I create directly from VS2017 templates that I do not edit myself at all, I run them as is. Empty, Web API and Web Application all still fail, not on NuGet but on building directly. Here is a screenshot showing exactly what I'm trying to explain.

If I am posting these in the wrong place perhaps you wouldn't mind giving me a small nudge in the correct area. Thanks.

![vtfs-buildfail](https://cloud.githubusercontent.com/assets/16159685/25094402/fb3ed304-2364-11e7-9d85-e3bfa9bbab1d.png)

**Comment by mbp** [comment]
It is now possible to pick NuGet 4 with NuGet Restore and NuGet Publisher. However, we still cannot pick version for NuGet Packager.

**Comment by alexmullans** [comment]
@mbp yep, we're still working on a set of updates to the Packager and Publisher tasks. In the meantime, [this workaround](https://github.com/Microsoft/vsts-tasks/issues/3756#issuecomment-285520122) will work for the Packager task.

@DanJ210 this is a good place to post, but unfortunately I'm not super-familiar with the Visual Studio Build task. I'm trying to find someone else internally who can help.

**Comment by channeladam** [comment]
Is there an ETA for the update to NuGet Packager?
This issue and having to do a workaround isn't a great story when I'm trying to convince a new customer to use VSTS....
Thanks

**Comment by alexmullans** [comment]
Spec is finished, work is starting this week. Most likely, it'll land in the deployment after next, so ~1 month from now.

**Comment by alexmullans** [comment]
This is fixed by #4397. We're releasing a new unified NuGet task that will use NuGet 4.0 by default but that can also pick up any version of NuGet provided by the new NuGet Tool Installer task.

**Comment by gbatterbee** [comment]
Glad to hear this has been fixed. What's the process of using this? Is it an update for VSTS to I need to wait? Is in in the market place? Many thanks.

**Comment by alexmullans** [comment]
The tasks will be deployed as part of the next deployment in 2ish weeks. You'll see a notification on your task that a new version is available.

**Comment by eriove** [comment]
Is this available for on premises TFS? If not, when will it be?

**Comment by alexmullans** [comment]
This work will land in TFS 2018.
